### ISSUE TITLE
How does a verifier correlate a presentation proposal with the proper presentation request?

### ISSUE THREAD

**Issue by brianorwhatever** [decentralized-identity/waci-presentation-exchange#99]
Here are the steps as I understand them with the solution I will work towards. Interested in hearing others thoughts.

1. The verifier creates a Presentation Request, which in turn creates a didcomm out of band invitation (no correlating factor between these two things)
2. The prover reads the invitation and sends a didcomm present-proof "proposal" message
3. The verifier receives the proposal and sends a didcomm present-proof "request" message including the PE definition
4. The prover receives the request and sends the submission as a didcomm present-proof "present" message

This sticking point is step 3 because the verifier has no way of knowing which presentation request is being proposed to. The simple solution would be for the verifier to save the invitation id onto the presentation request data model. Do we think this is the correct approach?

**Comment by agropper** [comment]
From a privacy perspective, 1 and 3 are separate requests. 1 is a public
and non-specific advertisement to potential provers. The advertisement
offers one or more schemas that a prover can choose to respond with. The
response, if any, is a choice of schema. 3 is a request for private
information or a ZKP.

Step 1 forces the prover to make some choices, thereby leaking some
information to the verifier. Step 3 is the actual authorization request
which, if approved, results in private information.

The establishment of a messaging link in step 2 might or might not identify
the prover (as subject).

Step 3 could have been step 1 if the advertisement of 1 actually held just
one schema. In that case, the prover might reply with a VC, a VP, a ZKP, an
access token to a VC or VP, or not reply at all.

- Adrian

**Comment by brianorwhatever** [comment]
Step 1 is entirely on the verifier side. The prove is not involved in the interaction whatsoever so they do not need to make any choices. Step 2 is where the prover pings the verifier telling them they propose sending  a presentation (without making any choices either). Step 3 is the first instance of anything related to VCs being sent across the didcomm channel. Yes, this is where the authorization request is included.

My question is specifically: how can the verifier know what presentation request they created the invitation for so that they can send this data across in step 3. The only way it can be done in the current specification is by linking the invitation and the request together internally

**Comment by agropper** [comment]
On Wed, Sep 8, 2021 at 7:57 PM Brian Richter ***@***.***>
wrote:

> Step 1 is entirely on the verifier side. The prove is not involved in the
> interaction whatsoever so they do not need to make any choices. Step 2 is
> where the prover pings the verifier telling them they propose sending a
> presentation (without making any choices either). Step 3 is the first
> instance of anything related to VCs being sent across the didcomm channel.
> Yes, this is where the authorization request is included.
>
I agree. 1 is just a verifier's advertisement and 1 and 2 together are a
request-response pair where the response might "leak" an identifier for the
entity that might later authorize release of a VC or VP to the verifier.
The leaked identifier (an endpoint or resolvable to and endpoint) might be
anonymous, pairwise pseudonymous (including use of a mediator), or
correlatable with the subject. The "real" authorization request is Step
3-4.

> My question is specifically: how can the verifier know what presentation
> request they created the invitation for so that they can send this data
> across in step 3. The only way it can be done in the current specification
> is by linking the invitation and the request together internally
>
My question, from a privacy perspective, is why engineer two (linked)
requests at all? A single request in the form of a verifier's
advertisement, can result in a VC-related response in a way that leaks no
data at all if the prover chooses to ignore the request. Because the prover
incurs a cost in processing requests, some advertisements can offer a
bounty to the prover to offset that cost.

In other words, I'm questioning the use of messaging in this way. It's like
emailing out a "survey" with a nonce and keeping a list of which nonce
paired with which email address so that if a recipient answers the "survey"
their email is known even if they tried to provide another identifier or
endpoint.

- Adrian

**Comment by brianorwhatever** [comment]
Ok I think I understand your concern a bit better. Using your example here what I'm trying to implement right now is on the verifier side in step 3 I need to know which survey to send to an email address.

As the verifier in step 3 I have received a proposal from a certain identifier (who i will send  to) so now I need to look up in my database of requests which request it is they are looking to fulfil (what i will send). The only way I can see to find the what is by having a link from request -> invitation.

Without this linkage there is no way for me as the verifier to know what the prover is trying to prove

**Comment by brianorwhatever** [comment]
I guess a similar question I should be asking is a Presentation Request unique to a prover? Your comments here bring me to thinking you believe the answer should be no however I don't think the current specification mentions anything about this

**Comment by brianorwhatever** [comment]
> My question, from a privacy perspective, is why engineer two (linked)
requests at all? A single request in the form of a verifier's
advertisement, can result in a VC-related response in a way that leaks no
data at all if the prover chooses to ignore the request.

So are you proposing to include the presentation request details into step 1? This might actually be a good idea as it would advertise what the verifier is looking for directly, giving the prover the opportunity to see what is being asked of them before even communicating with the verifier giving them the option to simply ignore the request. They would still need to go through steps 2 and 3 for the challenge / authentication so it wouldn't remove any of the steps.

I also have been wondering why we mention that the propose message is optional since I can't see any way for a verifier to know who to send a request to without it. I'll open a separate issue for that

**Comment by agropper** [comment]
Exactly! This is why a unified request advertisement is better from a
privacy engineering perspective. I'm wondering what the use-cases would be
where an advertisement followed by a request are necessary. Once we
understand that, we would be able to design the linkage between the
advertisement and the request.

- Adrian

**Comment by brentzundel** [comment]
addressed by PR #108, closing
